src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/QinQFlowSpec.groovy

```diff
@@ -0,0 +1,888 @@
+        def swP = topologyHelper.getAllNeighboringSwitchPairs().find {
+            [it.src, it.dst].every { sw ->
+                sw.dpId in allTraffGenSwitches*.dpId && northbound.getSwitchProperties(sw.dpId).multiTable
+            }
+        } ?: assumeTrue("No suiting switches found", false)
+
+        when: "Create a QinQ flow"
+        def flowWithQinQ = flowHelperV2.randomFlow(swP)
+        flowWithQinQ.source.innerVlanId = 234
+        flowWithQinQ.destination.innerVlanId = 432
+        flowHelperV2.addFlow(flowWithQinQ)
+
+        and: "Create a flow without QinQ"
+        def flowWithoutQinQ = flowHelperV2.randomFlow(swP)
+        flowWithoutQinQ.source.vlanId = 0
+        flowWithoutQinQ.source.innerVlanId = flowWithQinQ.source.vlanId
+        flowHelperV2.addFlow(flowWithoutQinQ)
+
+        then: "Both flows allow traffic"
+        def traffExam = traffExamProvider.get()
+        def examFlowWithtQinQ = new FlowTrafficExamBuilder(topology, traffExam)
+                .buildBidirectionalExam(flowHelperV2.toV1(flowWithQinQ), 1000, 5)
+        def examFlowWithoutQinQ = new FlowTrafficExamBuilder(topology, traffExam)
+                .buildBidirectionalExam(flowHelperV2.toV1(flowWithoutQinQ), 1000, 5)
+        withPool {
+            [examFlowWithtQinQ.forward, examFlowWithtQinQ.reverse,
+             examFlowWithoutQinQ.forward, examFlowWithoutQinQ.reverse].eachParallel { direction ->
+                def resources = traffExam.startExam(direction)
+                direction.setResources(resources)
+                assert traffExam.waitExam(direction).hasTraffic()
+            }
+        }
+
+        and: "Revert system to original state"
+        [flowWithQinQ, flowWithoutQinQ].each { flowHelperV2.deleteFlow(it.flowId) }
+    }
+
+    def "System detects conflict QinQ flows"() {
+        given: "Two switches with enabled multi table mode"
+        def swP = topologyHelper.getAllNeighboringSwitchPairs().find {
+            [it.src, it.dst].every { northbound.getSwitchProperties(it.dpId).multiTable }
+        } ?: assumeTrue("No suiting switches found", false)
+
+        when: "Create a first flow"
+        def flow1 = flowHelperV2.randomFlow(swP)
+        flow1.source.vlanId = 10
+        flow1.source.innerVlanId = 0
+        flowHelperV2.addFlow(flow1)
+
+        and: "Try to create a conflict flow"
+        def flow2 = flowHelperV2.randomFlow(swP)
+        flow2.source.vlanId = 0
+        flow2.source.innerVlanId = flow1.source.vlanId
+        flow2.source.portNumber = flow1.source.portNumber
+        northboundV2.addFlow(flow2)
+
+        then: "Human readable error is returned"
+        def exc = thrown(HttpClientErrorException)
+        exc.statusCode == HttpStatus.CONFLICT
+        exc.responseBodyAsString.to(MessageError).errorMessage == "Could not create flow"
+
+        and: "Revert system to original state"
+        flowHelperV2.deleteFlow(flow1.flowId)
+    }
```

Comment:
Well both flows in this test are actually non-qinq..And laconic test/step descriptions make it even more difficult to understand

Reply:
added 'where' block
```
where:
        vlan | innerVlan | conflictVlan | conflictInnerVlanId
        10   | 100       | 10           | 100
        10   | 0         | 0            | 10
```